File: morse.go
package main

import (
	"fmt"
	"math"
	"strings"
	"unicode"
)

const DitsInBeat = 8

type SequencerStep struct {
	DitDuration int
	Frequency   float64 // Frequency of the sine wave in Hz
}

func textToMorse(text string) string {
	// Morse code representation
	morseCodeDict := map[rune]string{
		'A': ".-", 'B': "-...", 'C': "-.-.", 'D': "-..", 'E': ".",
		'F': "..-.", 'G': "--.", 'H': "....", 'I': "..", 'J': ".---",
		'K': "-.-", 'L': ".-..", 'M': "--", 'N': "-.", 'O': "---",
		'P': ".--.", 'Q': "--.-", 'R': ".-.", 'S': "...", 'T': "-",
		'U': "..-", 'V': "...-", 'W': ".--", 'X': "-..-", 'Y': "-.--",
		'Z': "--..",

		'1': ".----", '2': "..---", '3': "...--", '4': "....-", '5': ".....",
		'6': "-....", '7': "--...", '8': "---..", '9': "----.", '0': "-----",

		',': "--..--", '.': ".-.-.-", '?': "..--..", '/': "-..-.", '-': "-....-",
		'(': "-.--.", ')': "-.--.-",

		' ': "_", // This is used to explicitly track the space between words
	}

	// Add two dummy blank spaces at the end of the more code, to make looping more intuitive
	text += "  "

	// Converting text to uppercase
	text = strings.ToUpper(text)

	// Translating each character to Morse code
	var morseCode []string
	for _, char := range text {
		morseCode = append(morseCode, morseCodeDict[char])
	}

	return strings.Join(morseCode, " ")
}

// GenerateSineWave generates a complete sine wave based on the SineWaveConfig
// If the duration doesn't match an even sine wave period, modify the duration.
func GenerateSineWave(frequency float64, sampleRate int, duration float64, fadeTime float64) []int16 {
	completeCyclesDuration := math.Round(frequency*duration) / frequency
	samplesCount := int(math.Round(float64(sampleRate) * completeCyclesDuration))
	sineWave := make([]int16, samplesCount)

	for i := 0; i < samplesCount; i++ {
		// Calculate the sine value for this sample
		sample := math.Sin(2 * math.Pi * frequency * float64(i) / float64(sampleRate))
		// Convert it to 16-bit integer range
		sineWave[i] = int16(sample * 32767)
	}

	fadeSamples := int(fadeTime * float64(sampleRate))
	// Fade in sample
	for i := 0; i < fadeSamples; i++ {
		fadeInMultiplier := float64(i) / float64(fadeSamples)
		sineWave[i] = int16(float64(sineWave[i]) * fadeInMultiplier)
	}

	// Fade out sample
	fadeoutOffset := len(sineWave) - fadeSamples
	for i := 0; i < fadeSamples; i++ {
		fadeOutMultiplier := float64(fadeSamples-i-1) / float64(fadeSamples)
		sineWave[i+fadeoutOffset] = int16(float64(sineWave[i+fadeoutOffset]) * fadeOutMultiplier)
	}

	return sineWave
}

// NoteToFrequency converts a musical note to its frequency.
// Note should be in the format "C4", "A4", "F#3", etc.
func NoteToFrequency(note string) (float64, error) {
	const ReferenceFrequencyA4 = 440.0
	const NotesInOctave = 12

	// Define the notes
	notes := map[string]int{
		"C": -9, "C#": -8, "D": -7, "D#": -6, "E": -5, "F": -4,
		"F#": -3, "G": -2, "G#": -1, "A": 0, "A#": 1, "B": 2,
	}

	note = strings.ToUpper(note)

	// Split the note into the note name and octave
	if len(note) < 2 {
		return 0, fmt.Errorf("invalid note: %s", note)
	}

	// Find the position where the octave starts (first digit)
	var octaveStart int
	for i, r := range note {
		if unicode.IsDigit(r) {
			octaveStart = i
			break
		}
	}

	if octaveStart == 0 {
		return 0, fmt.Errorf("invalid note: %s", note)
	}

	noteName := note[:octaveStart]
	octave, err := parseOctave(note[octaveStart:])
	if err != nil {
		return 0, err
	}

	// Calculate the position of the note
	position, ok := notes[noteName]
	if !ok {
		return 0, fmt.Errorf("invalid note: %s", note)
	}

	// Calculate the number of half steps from A4
	halfSteps := position + (octave-4)*NotesInOctave

	// Calculate the frequency
	frequency := ReferenceFrequencyA4 * math.Pow(2, float64(halfSteps)/12)

	return frequency, nil
}

// parseOctave parses the octave part of the note and checks for errors.
func parseOctave(octaveString string) (int, error) {
	const TopOctave = 8
	var octave int
	_, err := fmt.Sscanf(octaveString, "%d", &octave)
	if err != nil || octave < 0 || octave > TopOctave {
		return 0, fmt.Errorf("invalid octave: %s", octaveString)
	}
	return octave, nil
}

// Calculate a suitable dit length based on a tempo in BPM
func ditLengthForTempo(tempo float64) float64 {
	const SecondsInMinute = 60
	return SecondsInMinute / tempo / DitsInBeat
}

func main() {
	var err error
	tempo := 160.0
	note := "A5"
	frequency, err := NoteToFrequency(note)
	inputText := "stop radioactivity stop radioactivity stop radioactivity stop radioactivity"
	sampleRate := 44100
	ditLength := ditLengthForTempo(tempo)

	morseCode := textToMorse(inputText)

	var outputData []int16

	sequence := []SequencerStep{
		{8, 523.25},
		{8, 587.33},
		{8, 659.25},
		{8, 783.99},
		{8, 1046.50},
		{8, 1174.66},
		{8, 1318.51},
		{8, 1174.66},
		{8, 1046.50},
		{8, 783.99},
		{8, 659.25},
		{8, 587.33},

		{8, 523.25},
		{8, 587.33},
		{8, 659.25},
		{8, 783.99},
		{8, 1046.50},
		{8, 1174.66},
		{8, 1318.51},
		{8, 1174.66},
		{8, 1046.50},
		{8, 783.99},
		{8, 659.25},
		{8, 587.33},

		{8, 440.00},
		{8, 493.88},
		{8, 523.25},
		{8, 659.25},
		{8, 880.00},
		{8, 987.77},
		{8, 1046.50},
		{8, 987.77},
		{8, 880.00},
		{8, 659.25},
		{8, 523.25},
		{8, 493.88},

		{8, 440.00},
		{8, 493.88},
		{8, 523.25},
		{8, 659.25},
		{8, 880.00},
		{8, 987.77},
		{8, 1046.50},
		{8, 987.77},
		{8, 880.00},
		{8, 659.25},
		{8, 523.25},
		{8, 493.88},

		{8, 349.23},
		{8, 392.00},
		{8, 440.00},
		{8, 523.25},
		{8, 698.46},
		{8, 783.99},
		{8, 880.00},
		{8, 783.99},
		{8, 698.46},
		{8, 523.25},
		{8, 440.00},
		{8, 392.00},

		{8, 349.23},
		{8, 392.00},
		{8, 440.00},
		{8, 523.25},
		{8, 698.46},
		{8, 783.99},
		{8, 880.00},
		{8, 783.99},
		{8, 698.46},
		{8, 523.25},
		{8, 440.00},
		{8, 392.00},

		{999999999, 440},
	}

	ditCounter := 0
	sequenceAccumulator := 0
	sequenceStep := 0

	dotSample := GenerateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength, 0.005)
	dashSample := GenerateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength*3, 0.005)

	for _, runeValue := range morseCode {
		if ditCounter >= sequenceAccumulator+sequence[sequenceStep].DitDuration {
			sequenceAccumulator += sequence[sequenceStep].DitDuration
			sequenceStep++
			// Only generate new dot and dash samples on every new sequencer step
			dotSample = GenerateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength, 0.005)
			dashSample = GenerateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength*3, 0.005)
		}

		switch runeValue {
		case '.':
			outputData = append(outputData, dotSample...)
			expectedLength := int(float64(ditCounter+2) * (ditLength) * float64(sampleRate))
			outputData = append(outputData, make([]int16, expectedLength-len(outputData))...)
			ditCounter += 2
		case '-':
			outputData = append(outputData, dashSample...)
			expectedLength := int(float64(ditCounter+4) * (ditLength) * float64(sampleRate))
			outputData = append(outputData, make([]int16, expectedLength-len(outputData))...)
			ditCounter += 4
		case '_':
			expectedLength := int(float64(ditCounter+0) * (ditLength) * float64(sampleRate))
			outputData = append(outputData, make([]int16, expectedLength-len(outputData))...)
			ditCounter += 0
		case ' ':
			expectedLength := int(float64(ditCounter+4) * (ditLength) * float64(sampleRate))
			outputData = append(outputData, make([]int16, expectedLength-len(outputData))...)
			ditCounter += 4
		}

	}

	// Saving the data as a WAV file
	err = SaveAsWAV("output.wav", sampleRate, outputData)
	if err != nil {
		fmt.Println("Error saving WAV file:", err)
	} else {
		fmt.Println("WAV file saved successfully.")
	}

	fmt.Println(frequency)
	fmt.Println(ditLength)
}

File: morse.js
const fs = require('fs');

function textToMorse(text) {
    // Morse code representation
    const morseCodeDict = {
        'A': ".-", 'B': "-...", 'C': "-.-.", 'D': "-..", 'E': ".",
        'F': "..-.", 'G': "--.", 'H': "....", 'I': "..", 'J': ".---",
        'K': "-.-", 'L': ".-..", 'M': "--", 'N': "-.", 'O': "---",
        'P': ".--.", 'Q': "--.-", 'R': ".-.", 'S': "...", 'T': "-",
        'U': "..-", 'V': "...-", 'W': ".--", 'X': "-..-", 'Y': "-.--",
        'Z': "--..",

        '1': ".----", '2': "..---", '3': "...--", '4': "....-", '5': ".....",
        '6': "-....", '7': "--...", '8': "---..", '9': "----.", '0': "-----",

        ',': "--..--", '.': ".-.-.-", '?': "..--..", '/': "-..-.", '-': "-....-",
        '(': "-.--.", ')': "-.--.-",

        ' ': "_", // This is used to explicitly track the space between words
    };

    // Add two dummy blank spaces at the end of the Morse code, to make looping more intuitive
    text += "  ";

    // Converting text to uppercase
    text = text.toUpperCase();

    // Translating each character to Morse code
    const morseCode = [];
    for (const char of text) {
        if (morseCodeDict[char]) {
            morseCode.push(morseCodeDict[char]);
        }
    }

    return morseCode.join(" ");
}

function generateSineWave(frequency, sampleRate, duration, fadeTime) {
    const completeCyclesDuration = Math.round(frequency * duration) / frequency;
    const samplesCount = Math.round(sampleRate * completeCyclesDuration);
    const floatSineWave = new Array(samplesCount);
    const sineWave = new Array(samplesCount);

    for (let i = 0; i < samplesCount; i++) {
        // Calculate the sine value for this sample
        const sample = Math.sin((2 * Math.PI * frequency * i) / sampleRate);
        // Convert it to 16-bit integer range
        floatSineWave[i] = sample * 32767;
    }

    const fadeSamples = Math.round(fadeTime * sampleRate);
    // Fade in sample
    for (let i = 0; i < fadeSamples; i++) {
        const fadeInMultiplier = i / fadeSamples;
        floatSineWave[i] *= fadeInMultiplier;
    }

    // Fade out sample
    const fadeoutOffset = floatSineWave.length - fadeSamples;
    for (let i = 0; i < fadeSamples; i++) {
        const fadeOutMultiplier = (fadeSamples - i - 1) / fadeSamples;
        floatSineWave[i + fadeoutOffset] *= fadeOutMultiplier;
    }

    // Convert to 16-bit integers
    for (let i = 0; i < samplesCount; i++) {
        sineWave[i] = Math.round(floatSineWave[i]);
    }

    return sineWave;
}

// NoteToFrequency converts a musical note to its frequency.
// Note should be in the format "C4", "A4", "F#3", etc.
function noteToFrequency(note) {
    const referenceFrequencyA4 = 440.0;
    const notesInOctave = 12;

    // Define the notes
    const notes = {
        "C": -9, "C#": -8, "D": -7, "D#": -6, "E": -5, "F": -4,
        "F#": -3, "G": -2, "G#": -1, "A": 0, "A#": 1, "B": 2,
    };

    note = note.toUpperCase();

    // Split the note into the note name and octave
    if (note.length < 2) {
        throw new Error("invalid note: " + note);
    }

    // Find the position where the octave starts (first digit)
    let octaveStart = 0;
    for (let i = 0; i < note.length; i++) {
        if (!isNaN(note[i])) {
            octaveStart = i;
            break;
        }
    }

    if (octaveStart === 0) {
        throw new Error("invalid note: " + note);
    }

    const noteName = note.slice(0, octaveStart);
    const octave = parseOctave(note.slice(octaveStart));

    // Calculate the position of the note
    const position = notes[noteName];
    if (position === undefined) {
        throw new Error("invalid note: " + note);
    }

    // Calculate the number of half steps from A4
    const halfSteps = position + (octave - 4) * notesInOctave;

    // Calculate the frequency
    const frequency = referenceFrequencyA4 * Math.pow(2, halfSteps / 12);

    return frequency;
}

// parseOctave parses the octave part of the note and checks for errors.
function parseOctave(octaveString) {
    const topOctave = 8;
    const octave = parseInt(octaveString);

    if (isNaN(octave) || octave < 0 || octave > topOctave) {
        throw new Error("invalid octave: " + octaveString);
    }

    return octave;
}

function ditLengthForTempo(tempo) {
    const SecondsInMinute = 60;
    const DitsInBeat = 8;
    return SecondsInMinute / tempo / DitsInBeat;
}

class WAVHeader {
    constructor() {
        this.RiffTag = Buffer.from('RIFF');
        this.RiffSize = 0;
        this.WaveTag = Buffer.from('WAVE');
        this.FmtTag = Buffer.from('fmt ');
        this.FmtSize = 16; // PCM
        this.AudioFormat = 1; // Linear quantization
        this.NumChannels = 1; // Mono
        this.SampleRate = 0;
        this.ByteRate = 0;
        this.BlockAlign = 0;
        this.BitsPerSample = 0;
        this.DataTag = Buffer.from('data');
        this.DataSize = 0;
    }
}

function NewWAVHeader(sampleRate, bitsPerSample, dataLength) {
    const header = new WAVHeader();
    header.SampleRate = sampleRate;
    header.BitsPerSample = bitsPerSample;
    header.ByteRate = (sampleRate * bitsPerSample) / 8;
    header.BlockAlign = (bitsPerSample / 8);
    header.DataSize = dataLength * (bitsPerSample / 8);
    header.RiffSize = header.DataSize + 36;

    return header;
}

function saveAsWAV(filename, sampleRate, data) {
    const header = NewWAVHeader(sampleRate, 16, data.length);
    const headerSize = 44; // Size of WAV header

    // Create a buffer to store the WAV data
    const buffer = Buffer.alloc(headerSize + data.length * 2); // 16-bit audio (2 bytes per sample)

    // Write the WAV header to the buffer
    header.RiffSize = headerSize + data.length * 2 - 8;
    header.SampleRate = sampleRate;
    header.BitsPerSample = 16;
    header.ByteRate = (sampleRate * 16) / 8;
    header.BlockAlign = 2;
    header.DataSize = data.length * 2;
    header.FmtSize = 16;

    let offset = 0;
    offset += header.RiffTag.copy(buffer, offset);
    offset = buffer.writeUInt32LE(header.RiffSize, offset);
    offset += header.WaveTag.copy(buffer, offset);
    offset += header.FmtTag.copy(buffer, offset);
    offset = buffer.writeUInt32LE(header.FmtSize, offset);
    offset = buffer.writeUInt16LE(header.AudioFormat, offset);
    offset = buffer.writeUInt16LE(header.NumChannels, offset);
    offset = buffer.writeUInt32LE(header.SampleRate, offset);
    offset = buffer.writeUInt32LE(header.ByteRate, offset);
    offset = buffer.writeUInt16LE(header.BlockAlign, offset);
    offset = buffer.writeUInt16LE(header.BitsPerSample, offset);
    offset += header.DataTag.copy(buffer, offset);
    offset = buffer.writeUInt32LE(header.DataSize, offset);

    // Write the audio data to the buffer (16-bit little-endian)
    for (let i = 0; i < data.length; i++) {
        buffer.writeInt16LE(data[i], offset);
        offset += 2;
    }

    // Write the buffer contents to the WAV file
    try {
        fs.writeFileSync(filename, buffer);
        return null; // Success, no error
    } catch (err) {
        return err; // Error occurred
    }
}

function createMorseCodeAudioData(morseCode, sampleRate, tempo) {
    let outputData = [];

    const ditLength = ditLengthForTempo(tempo); // Assuming this function is defined elsewhere
    const note = "A5";
    const frequency = noteToFrequency(note); // Assuming this function is defined elsewhere
    const sequence = [{ DitDuration: 999999999, Frequency: frequency }];

    let ditCounter = 0;
    let sequenceAccumulator = 0;
    let sequenceStep = 0;

    let dotSample = generateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength, 0.005);
    let dashSample = generateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength * 3, 0.005);

    for (const runeValue of morseCode) {
        if (ditCounter >= sequenceAccumulator + sequence[sequenceStep].DitDuration) {
            sequenceAccumulator += sequence[sequenceStep].DitDuration;
            sequenceStep++;
            dotSample = generateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength, 0.005);
            dashSample = generateSineWave(sequence[sequenceStep].Frequency, sampleRate, ditLength * 3, 0.005);
        }

        switch (runeValue) {
            case '.':
                outputData = outputData.concat(dotSample);
                const expectedLength = Math.floor((ditCounter + 2) * ditLength * sampleRate);
                outputData = outputData.concat(new Array(expectedLength - outputData.length).fill(0));
                ditCounter += 2;
                break;
            case '-':
                outputData = outputData.concat(dashSample);
                const expectedLengthDash = Math.floor((ditCounter + 4) * ditLength * sampleRate);
                outputData = outputData.concat(new Array(expectedLengthDash - outputData.length).fill(0));
                ditCounter += 4;
                break;
            case '_':
                const expectedLengthSpace = Math.floor(ditCounter * ditLength * sampleRate);
                outputData = outputData.concat(new Array(expectedLengthSpace - outputData.length).fill(0));
                ditCounter += 0;
                break;
            case ' ':
                const expectedLengthSpaceBetweenWords = Math.floor((ditCounter + 4) * ditLength * sampleRate);
                outputData = outputData.concat(new Array(expectedLengthSpaceBetweenWords - outputData.length).fill(0));
                ditCounter += 4;
                break;
        }
    }
    return outputData;
}

function main() {
    try {
        const tempo = 120.0;
        const inputText = "stop radioactivity";
        const sampleRate = 44100;

        const morseCode = textToMorse(inputText); // Assuming this function is defined elsewhere
        const outputData = createMorseCodeAudioData(morseCode, sampleRate, tempo);

        const err = saveAsWAV("output.wav", sampleRate, outputData); // Assuming this function is defined elsewhere

        if (err) {
            console.error("Error saving WAV file:", err.message);
        } else {
            console.log("WAV file saved successfully.");
        }

    } catch (error) {
        console.error("Error:", error.message);
    }
}

// Call the main function to execute the code
main();
